### DISCUSSION TITLE
How to use multiple GPUs at once? Or in my case multiple GPU partitions?

### DISCUSSION THREAD

**Discussion by Flash-Of-Thunder** [neuralhydrology/neuralhydrology#62] [Q&A] [answered]
Hello again,

Haven't seen this one in the documentation so I figured I'd ask. I've had great success with running NH on a GPU partition on a cluster and am wondering if I can have the model run on multiple partitions. In the config file when I can set cuda: 0 or cuda: 1 etc. and it will recognize different partitions but only ever seems to be running on 1 (as far as I can tell). I tried some random syntax, e.g. cuda: 0,1, but only got error messages.

Cheers!

**Comment by kratzert** [comment, marked as answer]
We currently have no support for multi-GPU training and it is currently not high on our priority list, as we never had similar issues. In fact, usually we can train 3-4 models at the same time on a single 12GB GPU, as the VRAM footprint of our models is usually not really large. 

Out of curiosity: What are you currently working on that your model does not fit on a single GPU? Or what are the specs of your model/training setup?

**Reply by Flash-Of-Thunder** [reply]
This makes sense. I think by most standards my model is training quickly, my last run was 40 gauges/basins over 40 years finished an epoch in ~5 minutes (4 dynamic variables, 10 static). So 40 epochs was only 3.3 hours. I believe the it/s value is a measure of processing speed? It was about 5 it/s during training. The cluster uses NVIDIA's 2080 TIs but I don't know how many partitions each card is split into. 

My goal is to get NSE values for my basins similar or better to those published for existing models ([you & your colleagues summarized some SAC-SMA models and your own LSTMs](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2019WR026065)). So then if my LSTM models are performing well (at least above SAC-SMA) I wanted to start quantifying the impact of land development changes on streamflow. For this I was thinking of using a similar amount of basins ~500+, and so my computing time would go up, but still easily done over 1-2 days. 

Related to this, I was looking at my test period NSE values for these 40 basins and many of the basins perform well (NSE > .7) and a few weren't preforming well. Results similar to your paper I just linked, I was wondering if you had any thoughts on what might be the cause of those underperforming basins. As I recall, when I read that paper the source of poor performing basins wasn't discovered at the time of writing. I was looking at some of the graphs of obs vs simluated flow w/ prcp for my basins and it seemed like low-flow basins with flash-flood like behavior performed the worst. I was trying to think of the conditions and parameters that might cause this. One thought I had was: are there large and dry basins that can have localized rainfall to produce flow some times and not others. Then I was wondering what parameters might improve the LSTM and I was wondering something like splitting precipitation into prcp near the river and prcp >1 km from river?

**Reply by Flash-Of-Thunder** [reply]
Quick update, I switched to using cudalstm instead of customlstm and saw speeds increase by ~10x
